---
title: "CellChat_Prep"
author: "Elias Lai"
subtitle: ElMallah Lab, Duke University
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, eval = TRUE, comment = NA, 
                      message = FALSE, warning = FALSE,
                      fig.height = 6, fig.width = 9, fig.align = "center")
```

```{r load-packages, message = FALSE, warning = FALSE}
library(tidyverse)
library(dplyr)
library(conflicted)
library(Seurat)
library(CellChat)
library(patchwork)
library(reticulate)
library(ComplexHeatmap)
library(enrichR)
library(DropletUtils)
library(dplyr)
library(Matrix)
library(cowplot)
library(sctransform)
library(ggplot2)
library(SoupX)
library(here)
library(pheatmap)
library(biomaRt)
library(EnhancedVolcano)
library(slingshot)
library(RColorBrewer)
library(scales)
options(future.globals.maxSize= 8912896000000)
col_p = c("#F3E5C2","#EFB89A","#EF8F50", "#F5731D","#FA4602","#FE5100")
options(stringsAsFactors = FALSE)
conflict_prefer("union", "BiocGenerics")

# merge <- readRDS("output/merge.rds")
# merge.markers <- readRDS("output/merge.markers.rds")

# GAAKO_cellchat <- readRDS("output/GAAKO_cellchat.rds")
# Control_cellchat <- readRDS("output/Control_cellchat.rds")
# load(cellchat.RData)

```

```{r subset, message=FAlSE, warning=FALSE, eval = FALSE}
# subset based on control vs KO
GAAKO <- subset(x = merge, subset = condition == "GAA KO")
# saveRDS(GAAKO, file = "output/GAAKO.rds")
Control <- subset(x = merge, subset = condition == "Control")
saveRDS(Control, file = "output/Control.rds")

saveRDS(cellchat, file = "output/cellchat.rds")
```


```{r GAAKO_cellchat, message=FALSE, warning=FALSE, eval = FALSE}
options(stringsAsFactors = FALSE)
# load(url("https://ndownloader.figshare.com/files/25950872")) # example input for creating CellChat object

# Setting up GAAKO_cellchat object
GAAKO_cellchat <- createCellChat(object = GAAKO, group.by = "clusters")
rm(GAAKO)
levels(GAAKO_cellchat@idents) # show factor levels of the cell labels
groupSize <- as.numeric(table(GAAKO_cellchat@idents)) # number of cells in each cell group

# set the ligand-receptor interaction database
CellChatDB <- CellChatDB.mouse
CellChatDB.use <- CellChatDB # use all CellChatDB for cell-cell communication analysis (simply use the default CellChatDB)
GAAKO_cellchat@DB <- CellChatDB.use # set the used database in the object

# subset the expression data of signaling genes for saving computation cost
GAAKO_cellchat <- subsetData(GAAKO_cellchat) # This step is necessary even if using the whole database
future::plan("multiprocess", workers = 4) # do parallel
GAAKO_cellchat <- identifyOverExpressedGenes(GAAKO_cellchat)
GAAKO_cellchat <- identifyOverExpressedInteractions(GAAKO_cellchat)
GAAKO_cellchat <- projectData(GAAKO_cellchat, PPI.mouse) # project gene expression data onto PPI network (optional)

# compute probabilities
GAAKO_cellchat <- computeCommunProb(GAAKO_cellchat) # computing communication probability and infer cellular communication network
GAAKO_cellchat <- filterCommunication(GAAKO_cellchat, min.cells = 10) # Filter cell groups with few number of cells
GAAKO_cellchat <- computeCommunProbPathway(GAAKO_cellchat) # infer cell-cell communication at a signaling pathway level
# saveRDS(GAAKO_cellchat, file = "output/GAAKO_cellchat.rds")

# calculate aggregated cell-cellcommunication network
GAAKO_cellchat <- aggregateNet(GAAKO_cellchat) # calculate aggregated cell-cell communication network
groupSize <- as.numeric(table(GAAKO_cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(GAAKO_cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(GAAKO_cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
# control parameter edge.weight.max to compare edge weights between different networks
mat <- GAAKO_cellchat@net$weight
par(mfrow = c(3,4), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}

# Signaling pathways showing significant communications
View(GAAKO_cellchat@netP$pathways)

# Visualization of specific signaling pathways
pathways.show <- c("COLLAGEN") 
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells 
vertex.receiver = seq(1,4) # a numeric vector. 
netVisual_aggregate(GAAKO_cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver)
# Circle plot
par(mfrow=c(1,1))
netVisual_aggregate(GAAKO_cellchat, signaling = pathways.show, layout = "circle")
# Chord diagram
par(mfrow=c(1,1))
# netVisual_aggregate(GAAKO_cellchat, signaling = pathways.show, layout = "chord") # supposed to work
#> Note: The first link end is drawn out of sector 'Inflam. FIB'.
# Heatmap
par(mfrow=c(1,1))
netVisual_heatmap(GAAKO_cellchat, signaling = pathways.show, color.heatmap = "Reds")
#> Do heatmap based on a single object
# skipped some stuff from the tutorial

# compute contribution of each ligand-receptor pair to the overall signaling pathway and visualize cell-cell communication mediated by a single ligand-receptor pair
netAnalysis_contribution(GAAKO_cellchat, signaling = pathways.show)

# save plots of the all inferred network
# Access all the signaling pathways showing significant communications
pathways.show.all <- GAAKO_cellchat@netP$pathways
# check the order of cell identity to set suitable vertex.receiver
levels(GAAKO_cellchat@idents)
vertex.receiver = seq(1,4)
# setwd("CellChat_Plots")
# setwd("Control")
# for (i in 1:length(pathways.show.all)) {
#   # Visualize communication network associated with both signaling pathway and individual L-R pairs
#   netVisual(GAAKO_cellchat, signaling = pathways.show.all[i], 
#             vertex.receiver = vertex.receiver, layout = "hierarchy", 
#             out.format =  "png")
#   # Compute and visualize the contribution of each ligand-receptor pair to the overall signaling pathway
#   gg <- netAnalysis_contribution(GAAKO_cellchat, signaling = pathways.show.all[i])
#   ggsave(filename=paste0(pathways.show.all[i], "_L-R_contribution.pdf"), plot=gg, width = 24, height = 16, units = 'in', dpi = 300)
# }
# setwd("..")
# setwd("..")

# Visualize cell-cell communication mediated by multiple ligand-receptors or signaling pathways
# show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')
netVisual_bubble(GAAKO_cellchat, sources.use = 4, targets.use = c(5:11), remove.isolate = FALSE)
#> Comparing communications on a single object
# show all the significant interactions (L-R pairs) associated with certain signaling pathways
netVisual_bubble(GAAKO_cellchat, sources.use = 4, targets.use = c(5:11), signaling = c("LAMININ","COLLAGEN"), remove.isolate = FALSE)
#> Comparing communications on a single object

# Compute the network centrality scores
GAAKO_cellchat <- netAnalysis_computeCentrality(GAAKO_cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
# Visualize the computed centrality scores using heatmap, allowing ready identification of major signaling roles of cell groups
netAnalysis_signalingRole_network(GAAKO_cellchat, signaling = pathways.show, width = 8, height = 2.5, font.size = 10)

# Identify outgoing communication pattern of secreting cells
nPatterns = 3
GAAKO_cellchat <- identifyCommunicationPatterns(GAAKO_cellchat, pattern = c("outgoing"), k = nPatterns)
# river plot
# netAnalysis_river(GAAKO_cellchat, pattern = "outgoing")
# dot plot
netAnalysis_dot(GAAKO_cellchat, pattern = "outgoing")
# Identify incoming communication pattern of secreting cells
GAAKO_cellchat <- identifyCommunicationPatterns(GAAKO_cellchat, pattern = c("incoming"), k = nPatterns)
# river plot
# netAnalysis_river(GAAKO_cellchat, pattern = "incoming")
# dot plot
# netAnalysis_dot(GAAKO_cellchat, pattern = "incoming")

# Identify signaling groups based on their functional similarity
GAAKO_cellchat <- computeNetSimilarity(GAAKO_cellchat, type = "functional")
GAAKO_cellchat <- netEmbedding(GAAKO_cellchat, type = "functional")
GAAKO_cellchat <- netClustering(GAAKO_cellchat, type = "functional")
# Visualization in 2D-space
netVisual_embedding(GAAKO_cellchat, type = "functional", label.size = 3.5)
# netVisual_embeddingZoomIn(GAAKO_cellchat, type = "functional", nCol = 2)


# Identify signaling groups based on structure similarity
GAAKO_cellchat <- computeNetSimilarity(GAAKO_cellchat, type = "structural")
GAAKO_cellchat <- netEmbedding(GAAKO_cellchat, type = "structural")
GAAKO_cellchat <- netClustering(GAAKO_cellchat, type = "structural")
# Visualization in 2D-space
netVisual_embedding(GAAKO_cellchat, type = "structural", label.size = 3.5)
netVisual_embeddingZoomIn(GAAKO_cellchat, type = "structural", nCol = 2)

saveRDS(GAAKO_cellchat, file = "output/GAAKO_cellchat.rds")
```

```{r Control_cellchat, message=FALSE, warning=FALSE, eval = FALSE}
options(stringsAsFactors = FALSE)
# load(url("https://ndownloader.figshare.com/files/25950872")) # example input for creating CellChat object

# Setting up Control_cellchat object
Control_cellchat <- createCellChat(object = Control, group.by = "clusters")
rm(Control)
levels(Control_cellchat@idents) # show factor levels of the cell labels
groupSize <- as.numeric(table(Control_cellchat@idents)) # number of cells in each cell group

# set the ligand-receptor interaction database
CellChatDB <- CellChatDB.mouse
CellChatDB.use <- CellChatDB # use all CellChatDB for cell-cell communication analysis (simply use the default CellChatDB)
Control_cellchat@DB <- CellChatDB.use # set the used database in the object

# subset the expression data of signaling genes for saving computation cost
Control_cellchat <- subsetData(Control_cellchat) # This step is necessary even if using the whole database
future::plan("multiprocess", workers = 4) # do parallel
Control_cellchat <- identifyOverExpressedGenes(Control_cellchat)
Control_cellchat <- identifyOverExpressedInteractions(Control_cellchat)
Control_cellchat <- projectData(Control_cellchat, PPI.mouse) # project gene expression data onto PPI network (optional)

# compute probabilities
Control_cellchat <- computeCommunProb(Control_cellchat) # computing communication probability and infer cellular communication network
Control_cellchat <- filterCommunication(Control_cellchat, min.cells = 10) # Filter cell groups with few number of cells
Control_cellchat <- computeCommunProbPathway(Control_cellchat) # infer cell-cell communication at a signaling pathway level
# saveRDS(Control_cellchat, file = "output/Control_cellchat.rds")

# calculate aggregated cell-cellcommunication network
Control_cellchat <- aggregateNet(Control_cellchat) # calculate aggregated cell-cell communication network
groupSize <- as.numeric(table(Control_cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(Control_cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(Control_cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
# control parameter edge.weight.max to compare edge weights between different networks
mat <- Control_cellchat@net$weight
par(mfrow = c(3,4), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}

# Signaling pathways showing significant communications
View(Control_cellchat@netP$pathways)

# Visualization of specific signaling pathways
pathways.show <- c("COLLAGEN") 
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells 
vertex.receiver = seq(1,4) # a numeric vector. 
netVisual_aggregate(Control_cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver)
# Circle plot
par(mfrow=c(1,1))
netVisual_aggregate(Control_cellchat, signaling = pathways.show, layout = "circle")
# Chord diagram
par(mfrow=c(1,1))
# netVisual_aggregate(Control_cellchat, signaling = pathways.show, layout = "chord") # supposed to work
#> Note: The first link end is drawn out of sector 'Inflam. FIB'.
# Heatmap
par(mfrow=c(1,1))
netVisual_heatmap(Control_cellchat, signaling = pathways.show, color.heatmap = "Reds")
#> Do heatmap based on a single object
# skipped some stuff from the tutorial

# compute contribution of each ligand-receptor pair to the overall signaling pathway and visualize cell-cell communication mediated by a single ligand-receptor pair
netAnalysis_contribution(Control_cellchat, signaling = pathways.show)

# save plots of the all inferred network
# Access all the signaling pathways showing significant communications
pathways.show.all <- Control_cellchat@netP$pathways
# check the order of cell identity to set suitable vertex.receiver
levels(Control_cellchat@idents)
vertex.receiver = seq(1,4)
# setwd("CellChat_Plots")
# setwd("Control")
# for (i in 1:length(pathways.show.all)) {
#   # Visualize communication network associated with both signaling pathway and individual L-R pairs
#   netVisual(Control_cellchat, signaling = pathways.show.all[i], 
#             vertex.receiver = vertex.receiver, layout = "hierarchy", 
#             out.format =  "png")
#   # Compute and visualize the contribution of each ligand-receptor pair to the overall signaling pathway
#   gg <- netAnalysis_contribution(Control_cellchat, signaling = pathways.show.all[i])
#   ggsave(filename=paste0(pathways.show.all[i], "_L-R_contribution.pdf"), plot=gg, width = 24, height = 16, units = 'in', dpi = 300)
# }
# setwd("..")
# setwd("..")

# Visualize cell-cell communication mediated by multiple ligand-receptors or signaling pathways
# show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')
netVisual_bubble(Control_cellchat, sources.use = 4, targets.use = c(5:11), remove.isolate = FALSE)
#> Comparing communications on a single object
# show all the significant interactions (L-R pairs) associated with certain signaling pathways
netVisual_bubble(Control_cellchat, sources.use = 4, targets.use = c(5:11), signaling = c("LAMININ","COLLAGEN"), remove.isolate = FALSE)
#> Comparing communications on a single object

# Compute the network centrality scores
Control_cellchat <- netAnalysis_computeCentrality(Control_cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
# Visualize the computed centrality scores using heatmap, allowing ready identification of major signaling roles of cell groups
netAnalysis_signalingRole_network(Control_cellchat, signaling = pathways.show, width = 8, height = 2.5, font.size = 10)

# Identify outgoing communication pattern of secreting cells
nPatterns = 3
Control_cellchat <- identifyCommunicationPatterns(Control_cellchat, pattern = c("outgoing"), k = nPatterns)
# river plot
# netAnalysis_river(Control_cellchat, pattern = "outgoing")
# dot plot
netAnalysis_dot(Control_cellchat, pattern = "outgoing")
# Identify incoming communication pattern of secreting cells
Control_cellchat <- identifyCommunicationPatterns(Control_cellchat, pattern = c("incoming"), k = nPatterns)
# river plot
# netAnalysis_river(Control_cellchat, pattern = "incoming")
# dot plot
# netAnalysis_dot(Control_cellchat, pattern = "incoming")

# Identify signaling groups based on their functional similarity
Control_cellchat <- computeNetSimilarity(Control_cellchat, type = "functional")
Control_cellchat <- netEmbedding(Control_cellchat, type = "functional")
Control_cellchat <- netClustering(Control_cellchat, type = "functional")
# Visualization in 2D-space
netVisual_embedding(Control_cellchat, type = "functional", label.size = 3.5)
# netVisual_embeddingZoomIn(Control_cellchat, type = "functional", nCol = 2)


# Identify signaling groups based on structure similarity
Control_cellchat <- computeNetSimilarity(Control_cellchat, type = "structural")
Control_cellchat <- netEmbedding(Control_cellchat, type = "structural")
Control_cellchat <- netClustering(Control_cellchat, type = "structural")
# Visualization in 2D-space
netVisual_embedding(Control_cellchat, type = "structural", label.size = 3.5)
netVisual_embeddingZoomIn(Control_cellchat, type = "structural", nCol = 2)

saveRDS(Control_cellchat, file = "output/Control_cellchat.rds")
```

```{r comparison, message=FALSE, warning=FALSE, eval = FALSE}
# cd into directory for figures
# GAAKO_cellchat <- readRDS("output/GAAKO_cellchat.rds")
# Control_cellchat <- readRDS("output/Control_cellchat.rds")
setwd("CellChat_Plots")
setwd("Comparison")
rm(merge)

# merge datasets
object.list <- list(GAAKO = GAAKO_cellchat, Control = Control_cellchat)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))
rm(GAAKO_cellchat, Control_cellchat)

# compare interactions
gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")
gg1 + gg2

# Differential number of interactions or interaction strength among different cell populations
par(mfrow = c(1,2), xpd=TRUE)
netVisual_diffInteraction(cellchat, weight.scale = T)
netVisual_diffInteraction(cellchat, weight.scale = T, measure = "weight")

# Differential number of interactions with heatmap
gg1 <- netVisual_heatmap(cellchat)
gg2 <- netVisual_heatmap(cellchat, measure = "weight")
gg1 + gg2

# Compare outgoing and incoming interaction strength in 2D space
num.link <- sapply(object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
weight.MinMax <- c(min(num.link), max(num.link)) # control the dot size in the different datasets
gg <- list()
for (i in 1:length(object.list)) {
  gg[[i]] <- netAnalysis_signalingRole_scatter(object.list[[i]], title = names(object.list)[i], weight.MinMax = weight.MinMax)
}
patchwork::wrap_plots(plots = gg)

# Identify signaling groups based on their functional similarity
cellchat <- computeNetSimilarityPairwise(cellchat, type = "functional")
cellchat <- netEmbedding(cellchat, type = "functional")
cellchat <- netClustering(cellchat, type = "functional")
# Visualization in 2D-space
netVisual_embeddingPairwise(cellchat, type = "functional", label.size = 3.5)
# netVisual_embeddingZoomIn(cellchat, type = "functional", nCol = 2)

# Identify signaling groups based on structure similarity
cellchat <- computeNetSimilarityPairwise(cellchat, type = "structural")
cellchat <- netEmbedding(cellchat, type = "structural")
cellchat <- netClustering(cellchat, type = "structural")
# Visualization in 2D-space
netVisual_embeddingPairwise(cellchat, type = "structural", label.size = 3.5)
netVisual_embeddingPairwiseZoomIn(cellchat, type = "structural", nCol = 2)

# Compute and visualize the pathway distance in the learned joint manifold
rankSimilarity(cellchat, type = "functional")

# Compare the overall information flow of each signaling pathway
gg1 <- rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE)
gg2 <- rankNet(cellchat, mode = "comparison", stacked = F, do.stat = TRUE)
gg1 + gg2

# Compare outgoing (or incoming) signaling associated with each cell population
i = 1
# combining all the identified signaling pathways from different datasets 
pathway.union <- union(object.list[[i]]@netP$pathways, object.list[[i+1]]@netP$pathways)
ht1 = netAnalysis_signalingRole_heatmap(object.list[[i]], pattern = "outgoing", signaling = pathway.union, title = names(object.list)[i], width = 5, height = 6)
ht2 = netAnalysis_signalingRole_heatmap(object.list[[i+1]], pattern = "outgoing", signaling = pathway.union, title = names(object.list)[i+1], width = 5, height = 6)
draw(ht1 + ht2, ht_gap = unit(0.5, "cm"))
ht1 = netAnalysis_signalingRole_heatmap(object.list[[i]], pattern = "incoming", signaling = pathway.union, title = names(object.list)[i], width = 5, height = 6, color.heatmap = "GnBu")
ht2 = netAnalysis_signalingRole_heatmap(object.list[[i+1]], pattern = "incoming", signaling = pathway.union, title = names(object.list)[i+1], width = 5, height = 6, color.heatmap = "GnBu")
draw(ht1 + ht2, ht_gap = unit(0.5, "cm"))
ht1 = netAnalysis_signalingRole_heatmap(object.list[[i]], pattern = "all", signaling = pathway.union, title = names(object.list)[i], width = 5, height = 6, color.heatmap = "OrRd")
ht2 = netAnalysis_signalingRole_heatmap(object.list[[i+1]], pattern = "all", signaling = pathway.union, title = names(object.list)[i+1], width = 5, height = 6, color.heatmap = "OrRd")
draw(ht1 + ht2, ht_gap = unit(0.5, "cm"))

# Part III: Identify the upgulated and down-regulated signaling ligand-receptor pairs
# Identify dysfunctional signaling by comparing the communication probabities
netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11),  comparison = c(1, 2), angle.x = 45)

# identify the upgulated (increased) and down-regulated (decreased) signaling ligand-receptor pairs in one dataset compared to the other dataset
gg1 <- netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11),  comparison = c(1, 2), max.dataset = 2, title.name = "Increased signaling in LS", angle.x = 45, remove.isolate = T)
gg2 <- netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11),  comparison = c(1, 2), max.dataset = 1, title.name = "Decreased signaling in LS", angle.x = 45, remove.isolate = T)
gg1 + gg2

# Identify dysfunctional signaling by using differential expression analysis
# define a positive dataset, i.e., the dataset with positive fold change against the other dataset
pos.dataset = "Control"
# define a char name used for storing the results of differential expression analysis
features.name = pos.dataset
# perform differential expression analysis
cellchat <- identifyOverExpressedGenes(cellchat, group.dataset = "datasets", pos.dataset = pos.dataset, features.name = features.name, only.pos = FALSE, thresh.pc = 0.1, thresh.fc = 0.1, thresh.p = 1)
# map the results of differential expression analysis onto the inferred cell-cell communications to easily manage/subset the ligand-receptor pairs of interest
net <- netMappingDEG(cellchat, features.name = features.name)
# extract the ligand-receptor pairs with upregulated ligands in LS
net.up <- subsetCommunication(cellchat, net = net, datasets = "Control",ligand.logFC = 0.2, receptor.logFC = NULL)
# extract the ligand-receptor pairs with upregulated ligands and upregulated recetptors in NL, i.e.,downregulated in LS
net.down <- subsetCommunication(cellchat, net = net, datasets = "GAAKO",ligand.logFC = -0.1, receptor.logFC = -0.1)

# further deconvolution to obtain the individual signaling genes
gene.up <- extractGeneSubsetFromPair(net.up, cellchat)
gene.down <- extractGeneSubsetFromPair(net.down, cellchat)

# visualize the upgulated and down-regulated signaling ligand-receptor pairs using bubble plot or chord diagram
pairLR.use.up = net.up[, "interaction_name", drop = F]
gg1 <- netVisual_bubble(cellchat, pairLR.use = pairLR.use.up, sources.use = 4, targets.use = c(5:11), comparison = c(1, 2),  angle.x = 90, remove.isolate = T,title.name = paste0("Up-regulated signaling in ", names(object.list)[2]))
pairLR.use.down = net.down[, "interaction_name", drop = F]
gg2 <- netVisual_bubble(cellchat, pairLR.use = pairLR.use.down, sources.use = 4, targets.use = c(5:11), comparison = c(1, 2),  angle.x = 90, remove.isolate = T,title.name = paste0("Down-regulated signaling in ", names(object.list)[2]))
gg1 + gg2

# Visualize the upgulated and down-regulated signaling ligand-receptor pairs using Chord diagram
par(mfrow = c(1,2), xpd=TRUE)
netVisual_chord_gene(object.list[[2]], sources.use = 4, targets.use = c(5:11), slot.name = 'net', net = net.up, lab.cex = 0.8, small.gap = 3.5, title.name = paste0("Up-regulated signaling in ", names(object.list)[2]))
netVisual_chord_gene(object.list[[1]], sources.use = 4, targets.use = c(5:11), slot.name = 'net', net = net.down, lab.cex = 0.8, small.gap = 3.5, title.name = paste0("Down-regulated signaling in ", names(object.list)[2]))

# Part IV: Visually compare cell-cell communication using Hierarchy plot, Circle plot or Chord diagram
# visualize the cell-cell communication network using Hierarchy plot, Circle plot or Chord diagram
pathways.show <- c("COLLAGEN") 
weight.max <- getMaxWeight(object.list, slot.name = c("netP"), attribute = pathways.show) # control the edge weights across different datasets
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list[[i]], signaling = pathways.show, layout = "circle", edge.weight.max = weight.max[1], edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list)[i]))
}
# Heatmap
pathways.show <- c("COLLAGEN") 
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(object.list)) {
  ht[[i]] <- netVisual_heatmap(object.list[[i]], signaling = pathways.show, color.heatmap = "Reds",title.name = paste(pathways.show, "signaling ",names(object.list)[i]))
}
ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))
# Chord diagram
pathways.show <- c("COLLAGEN") 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list[[i]], signaling = pathways.show, layout = "chord", signaling.name = paste(pathways.show, names(object.list)[i]))
}
# Chord diagram
group.cellType <- c(rep("TypeIPneumocyte", 4), rep("Type2Pneumocyte", 4), rep("Goblet+Club+Basal", 4)) # grouping cell clusters into fibroblast, DC and TC cells
names(group.cellType) <- levels(object.list[[1]]@idents)
pathways.show <- c("COLLAGEN") 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_chord_cell(object.list[[i]], signaling = pathways.show, group = group.cellType, title.name = paste0(pathways.show, " signaling network - ", names(object.list)[i]))
}
par(mfrow = c(1, 2), xpd=TRUE)
# compare all the interactions sending from Inflam.FIB to DC cells
for (i in 1:length(object.list)) {
  netVisual_chord_gene(object.list[[i]], sources.use = 4, targets.use = c(5:8), lab.cex = 0.5, title.name = paste0("Signaling from Inflam.FIB - ", names(object.list)[i]))
}
# compare all the interactions sending from fibroblast to inflamatory immune cells
par(mfrow = c(1, 2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_chord_gene(object.list[[i]], sources.use = c(1,2, 3, 4), targets.use = c(8,10),  title.name = paste0("Signaling received by Inflam.DC and .TC - ", names(object.list)[i]), legend.pos.x = 10)
}
# show all the significant signaling pathways from fibroblast to immune cells
par(mfrow = c(1, 2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_chord_gene(object.list[[i]], sources.use = c(1,2,3,4), targets.use = c(5:11),slot.name = "netP", title.name = paste0("Signaling pathways sending from fibroblast - ", names(object.list)[i]), legend.pos.x = 10)
}

# plot the gene expression distribution of signaling genes related to L-R pairs or signaling pathway
cellchat@meta$datasets = factor(cellchat@meta$datasets, levels = c("Control", "GAAKO")) # set factor level
plotGeneExpression(cellchat, signaling = "COLLAGEN", split.by = "datasets", colors.ggplot = T)



saveRDS(cellchat, file = "output/cellchat.rds")


```
