---
title: "CellChat_Plots"
author: "Elias Lai"
subtitle: ElMallah Lab, Duke University
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, eval = TRUE, comment = NA, 
                      message = FALSE, warning = FALSE,
                      fig.height = 6, fig.width = 9, fig.align = "center")
```

```{r load-packages, message = FALSE, warning = FALSE}
library(tidyverse)
library(dplyr)
library(conflicted)
library(Seurat)
library(CellChat)
library(patchwork)
library(reticulate)
library(ComplexHeatmap)
library(enrichR)
library(DropletUtils)
library(dplyr)
library(Matrix)
library(cowplot)
library(sctransform)
library(ggplot2)
library(SoupX)
library(here)
library(pheatmap)
library(biomaRt)
library(EnhancedVolcano)
library(slingshot)
library(RColorBrewer)
library(scales)
options(future.globals.maxSize= 8912896000000)
col_p = c("#F3E5C2","#EFB89A","#EF8F50", "#F5731D","#FA4602","#FE5100")
options(stringsAsFactors = FALSE)
conflict_prefer("union", "BiocGenerics")

merge <- readRDS("output/merge.rds")
merge.markers <- readRDS("output/merge.markers.rds")

# GAAKO_cellchat <- readRDS("output/GAAKO_cellchat.rds")
# Control_cellchat <- readRDS("output/Control_cellchat.rds")
# load(cellchat.RData)

```



```{r cellchat-plots, message=FALSE, warning=FALSE, fig.show='hide'}
# setwd("CellChat_Plots")
# setwd("Comparison")
GAAKO_cellchat <- readRDS("output/GAAKO_cellchat.rds")
Control_cellchat <- readRDS("output/Control_cellchat.rds")
cellchat <- readRDS("output/cellchat.rds")
object.list <- list(GAAKO = GAAKO_cellchat, Control = Control_cellchat)
rm(GAAKO_cellchat, Control_cellchat)

# compare interactions
gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")
gg1 + gg2

# Differential number of interactions or interaction strength among different cell populations
par(mfrow = c(1,2), xpd=TRUE)
netVisual_diffInteraction(cellchat, weight.scale = T)
netVisual_diffInteraction(cellchat, weight.scale = T, measure = "weight")

# Differential number of interactions with heatmap
gg1 <- netVisual_heatmap(cellchat)
gg2 <- netVisual_heatmap(cellchat, measure = "weight")
gg1 + gg2

# Compare outgoing and incoming interaction strength in 2D space
num.link <- sapply(object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
weight.MinMax <- c(min(num.link), max(num.link)) # control the dot size in the different datasets
gg <- list()
for (i in 1:length(object.list)) {
  gg[[i]] <- netAnalysis_signalingRole_scatter(object.list[[i]], title = names(object.list)[i], weight.MinMax = weight.MinMax)
}
patchwork::wrap_plots(plots = gg)

# Visualization of signaling groups based on their functional similarity in 2D-space
netVisual_embeddingPairwise(cellchat, type = "functional", label.size = 3.5)
# netVisual_embeddingZoomIn(cellchat, type = "functional", nCol = 2)

# Visualization of signaling groups based on structure similarity in 2D-space
netVisual_embeddingPairwise(cellchat, type = "structural", label.size = 3.5)
netVisual_embeddingPairwiseZoomIn(cellchat, type = "structural", nCol = 2)

# Compute and visualize the pathway distance in the learned joint manifold
rankSimilarity(cellchat, type = "functional")

# Compare the overall information flow of each signaling pathway
gg1 <- rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE)
gg2 <- rankNet(cellchat, mode = "comparison", stacked = F, do.stat = TRUE)
gg1 + gg2

# Compare outgoing (or incoming) signaling associated with each cell population
i = 1
# combining all the identified signaling pathways from different datasets 
pathway.union <- union(object.list[[i]]@netP$pathways, object.list[[i+1]]@netP$pathways)
ht1 = netAnalysis_signalingRole_heatmap(object.list[[i]], pattern = "outgoing", signaling = pathway.union, title = names(object.list)[i], width = 5, height = 20)
ht2 = netAnalysis_signalingRole_heatmap(object.list[[i+1]], pattern = "outgoing", signaling = pathway.union, title = names(object.list)[i+1], width = 5, height = 20)
draw(ht1 + ht2, ht_gap = unit(0.5, "cm"))
ht1 = netAnalysis_signalingRole_heatmap(object.list[[i]], pattern = "incoming", signaling = pathway.union, title = names(object.list)[i], width = 5, height = 20, color.heatmap = "GnBu")
ht2 = netAnalysis_signalingRole_heatmap(object.list[[i+1]], pattern = "incoming", signaling = pathway.union, title = names(object.list)[i+1], width = 5, height = 20, color.heatmap = "GnBu")
draw(ht1 + ht2, ht_gap = unit(0.5, "cm"))
ht1 = netAnalysis_signalingRole_heatmap(object.list[[i]], pattern = "all", signaling = pathway.union, title = names(object.list)[i], width = 5, height = 20, color.heatmap = "OrRd")
ht2 = netAnalysis_signalingRole_heatmap(object.list[[i+1]], pattern = "all", signaling = pathway.union, title = names(object.list)[i+1], width = 5, height = 20, color.heatmap = "OrRd")
draw(ht1 + ht2, ht_gap = unit(0.5, "cm"))

# Part III: Identify the upgulated and down-regulated signaling ligand-receptor pairs
# Identify dysfunctional signaling by comparing the communication probabities
netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11),  comparison = c(1, 2), angle.x = 45, font.size = 12)

 # identify the upgulated (increased) and down-regulated (decreased) signaling ligand-receptor pairs in one dataset compared to the other dataset
gg1 <- netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11),  comparison = c(1, 2), max.dataset = 1, title.name = "Increased signaling in GAAKO", angle.x = 45, remove.isolate = T, font.size = 17)
gg2 <- netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11),  comparison = c(1, 2), max.dataset = 2, title.name = "Decreased signaling in GAAKO", angle.x = 45, remove.isolate = T, font.size = 14)
gg1
gg2

# Identify dysfunctional signaling by using differential expression analysis
# define a positive dataset, i.e., the dataset with positive fold change against the other dataset
pos.dataset = "GAAKO"
# define a char name used for storing the results of differential expression analysis
features.name = pos.dataset
# perform differential expression analysis
cellchat <- identifyOverExpressedGenes(cellchat, group.dataset = "datasets", pos.dataset = pos.dataset, features.name = features.name, only.pos = FALSE, thresh.pc = 0.1, thresh.fc = 0.1, thresh.p = 1)
# map the results of differential expression analysis onto the inferred cell-cell communications to easily manage/subset the ligand-receptor pairs of interest
net <- netMappingDEG(cellchat, features.name = features.name)
# extract the ligand-receptor pairs with upregulated ligands in LS
net.up <- subsetCommunication(cellchat, net = net, datasets = "GAAKO",ligand.logFC = 0.2, receptor.logFC = NULL)
# extract the ligand-receptor pairs with upregulated ligands and upregulated recetptors in NL, i.e.,downregulated in LS
net.down <- subsetCommunication(cellchat, net = net, datasets = "Control",ligand.logFC = -0.1, receptor.logFC = -0.1)

# further deconvolution to obtain the individual signaling genes
gene.up <- extractGeneSubsetFromPair(net.up, cellchat)
gene.down <- extractGeneSubsetFromPair(net.down, cellchat)

# visualize the upgulated and down-regulated signaling ligand-receptor pairs using bubble plot or chord diagram
pairLR.use.up = net.up[, "interaction_name", drop = F]
gg1 <- netVisual_bubble(cellchat, pairLR.use = pairLR.use.up, sources.use = 4, targets.use = c(5:11), comparison = c(1, 2),  angle.x = 45, remove.isolate = T,title.name = paste0("Up-regulated signaling in ", names(object.list)[1]))
pairLR.use.down = net.down[, "interaction_name", drop = F]
gg2 <- netVisual_bubble(cellchat, pairLR.use = pairLR.use.down, sources.use = 4, targets.use = c(5:11), comparison = c(1, 2),  angle.x = 45, remove.isolate = T,title.name = paste0("Down-regulated signaling in ", names(object.list)[1]))
gg1
gg2

# Visualize the upgulated and down-regulated signaling ligand-receptor pairs using Chord diagram
# par(mfrow = c(1,2), xpd=TRUE)
# netVisual_chord_gene(object.list[[2]], sources.use = 4, targets.use = c(5:11), slot.name = 'net', net = net.up, lab.cex = 0.8, small.gap = 3.5, title.name = paste0("Up-regulated signaling in ", names(object.list)[2]))
# netVisual_chord_gene(object.list[[1]], sources.use = 4, targets.use = c(5:11), slot.name = 'net', net = net.down, lab.cex = 0.8, small.gap = 3.5, title.name = paste0("Down-regulated signaling in ", names(object.list)[2]))

# Part IV: Visually compare cell-cell communication using Hierarchy plot, Circle plot or Chord diagram
# visualize the cell-cell communication network using Hierarchy plot, Circle plot or Chord diagram
pathways.show <- c("BMP") 
weight.max <- getMaxWeight(object.list, slot.name = c("netP"), attribute = pathways.show) # control the edge weights across different datasets
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list[[i]], signaling = pathways.show, layout = "circle", edge.weight.max = weight.max[1], edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list)[i]))
}
# Heatmap
pathways.show <- c("BMP") 
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(object.list)) {
  ht[[i]] <- netVisual_heatmap(object.list[[i]], signaling = pathways.show, color.heatmap = "Reds",title.name = paste(pathways.show, "signaling ",names(object.list)[i]))
}
ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"), row_title_gp = gpar(fontsize = 25), column_title_gp = gpar(fontsize = 25))
# Chord diagram
pathways.show <- c("BMP") 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list[[i]], signaling = pathways.show, layout = "chord", signaling.name = paste(pathways.show, names(object.list)[i]))
}
# # Chord diagram
# group.cellType <- c(rep("TypeIPneumocyte", 4), rep("Type2Pneumocyte", 4), rep("Goblet+Club+Basal", 4)) # grouping cell clusters into fibroblast, DC and TC cells
# names(group.cellType) <- levels(object.list[[1]]@idents)
# pathways.show <- c("BMP") 
# par(mfrow = c(1,2), xpd=TRUE)
# for (i in 1:length(object.list)) {
#   netVisual_chord_cell(object.list[[i]], signaling = pathways.show, group = group.cellType, title.name = paste0(pathways.show, " signaling network - ", names(object.list)[i]))
# }
par(mfrow = c(1, 2), xpd=TRUE)
# compare all the interactions sending from Inflam.FIB to DC cells
for (i in 1:length(object.list)) {
  netVisual_chord_gene(object.list[[i]], sources.use = 4, targets.use = c(5:8), lab.cex = 0.5, title.name = paste0("Signaling from Inflam.FIB - ", names(object.list)[i]))
}
# compare all the interactions sending from fibroblast to inflamatory immune cells
par(mfrow = c(1, 2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_chord_gene(object.list[[i]], sources.use = c(1,2, 3, 4), targets.use = c(8,10),  title.name = paste0("Signaling received by Inflam.DC and .TC - ", names(object.list)[i]), legend.pos.x = 10)
}
# show all the significant signaling pathways from fibroblast to immune cells
par(mfrow = c(1, 2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_chord_gene(object.list[[i]], sources.use = c(1,2,3,4), targets.use = c(5:11),slot.name = "netP", title.name = paste0("Signaling pathways sending from fibroblast - ", names(object.list)[i]), legend.pos.x = 10)
}

# plot the gene expression distribution of signaling genes related to L-R pairs or signaling pathway
cellchat@meta$datasets = factor(cellchat@meta$datasets, levels = c("Control", "GAAKO")) # set factor level
plotGeneExpression(cellchat, signaling = "BMP", split.by = "datasets", colors.ggplot = T)

```

```{r pathway1_specific_heatmaps, message=FALSE, warning=FALSE}
# Identified pathways1
identified_pathways1 <- c("EPHB", "CHEMERIN", "SEMA6", "CDH", "AGRN", "VTN", "BMP", "ANGPT", "SEMA7", "GALECTIN", "KIT", "ncWNT", "PERIOSTIN", "CDH1", "VISFATIN", "EGF", "PTPRM", "CD34")

# Heatmap
k = 1
for (k in 1:length(identified_pathways1)) {
  pathways.show <- c(identified_pathways1[k]) 
  par(mfrow = c(1,2), xpd=TRUE)
  ht <- list()
  for (i in 1:length(object.list)) {
    ht[[i]] <- netVisual_heatmap(object.list[[i]], signaling = pathways.show, color.heatmap = "Reds",title.name = paste(pathways.show, "signaling ",names(object.list)[i]))
  }
  ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"), row_title_gp = gpar(fontsize = 25), column_title_gp = gpar(fontsize = 25))
}
```


```{r}

identified_pathways2 <- c("GALECTIN", "KIT", "ncWNT", "PERIOSTIN", "CDH1", "VISFATIN", "EGF", "PTPRM", "CD34")

pathways.show <- c("CD34") 
  par(mfrow = c(1,2), xpd=TRUE)
  ht <- list()
  for (i in 1:length(object.list)) {
    ht[[i]] <- netVisual_heatmap(object.list[[i]], signaling = pathways.show, color.heatmap = "Reds",title.name = paste(pathways.show, "signaling ",names(object.list)[i]))
  }
  ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"), row_title_gp = gpar(fontsize = 25), column_title_gp = gpar(fontsize = 25))
```


```{r UMAPS, message=FALSE, warning=FALSE}
gg1 <- FeaturePlot(merge, c("Col4a1"))
gg2 <- DimPlot(merge, reduction = "umap", label = F,group.by = "condition")
gg3 <- DimPlot(merge, reduction = "umap", label = TRUE, pt.size = 0.001) + NoLegend()
gg1 + gg2 + gg3
```








